In [ ]:
pip install pingouin

In [ ]:
pip install pingouin --upgrade

# Import libraries

In [1]:
import pandas as pd
import numpy as np

# Load data

In [3]:
# Creating DataFrame from the provided data
data = {
    'plant_var': ['A']*10 + ['B']*10 + ['C']*10 + ['D']*10,
    'height': [
        17.997, 17.362, 18.148, 19.023, 17.266, 17.266, 19.079, 18.267, 17.031, 18.043,  # Group A
        17.037, 17.034, 17.742, 15.587, 15.775, 16.938, 16.487, 17.814, 16.592, 16.088,  # Group B
        18.966, 17.274, 17.568, 16.075, 16.956, 17.611, 16.349, 17.876, 16.899, 17.208,  # Group C
        16.898, 19.352, 17.487, 16.442, 18.323, 16.279, 17.709, 15.540, 16.172, 17.697   # Group D
    ],
    'canopy_vol': [
        0.71, 0.81, 0.92, 0.61, 0.73, 0.77, 0.85, 0.65, 0.98, 0.82,  # Group A
        0.41, 0.56, 0.59, 0.68, 0.69, 0.67, 0.60, 0.62, 0.53, 0.61,  # Group B
        0.11, 0.51, 0.13, 0.23, 0.35, 0.21, 0.15, 0.48, 0.30, 0.33,  # Group C
        0.31, 0.51, 0.56, 0.59, 0.42, 0.35, 0.50, 0.52, 0.55, 0.54   # Group D
    ]
}
df = pd.DataFrame(data)

**3.6b Expectation and Variance-Covariance Matrix**

In [5]:
# Mean (Expectation)
mean_values = df[['height', 'canopy_vol']].mean()
print("Mean (Expectation):")
print(mean_values)

# Covariance matrix
cov_matrix = df[['height', 'canopy_vol']].cov()
print("\nCovariance Matrix:")
print(cov_matrix)

Mean (Expectation):
height        17.281425
canopy_vol     0.536500
dtype: float64

Covariance Matrix:
              height  canopy_vol
height      0.907910    0.020464
canopy_vol  0.020464    0.045521


**3.6c Correlation Coefficient and Correlation Matrix**

In [7]:
# Correlation coefficient between height and canopy_vol
correlation = df['height'].corr(df['canopy_vol'])
print("\nCorrelation Coefficient between Height and Canopy Volume:", correlation)

# Correlation matrix
correlation_matrix = df[['height', 'canopy_vol']].corr()
print("\nCorrelation Matrix:")
print(correlation_matrix)


Correlation Coefficient between Height and Canopy Volume: 0.10065961851909648

Correlation Matrix:
             height  canopy_vol
height      1.00000     0.10066
canopy_vol  0.10066     1.00000


# Joint, Marginal and Conditional Probability Distributions

**3.6d Joint Probability Density Function (PDF)**

In [9]:
from scipy.stats import multivariate_normal

# Assuming height and canopy_vol follow a bivariate normal distribution
mean_vector = df[['height', 'canopy_vol']].mean().values
cov_matrix = df[['height', 'canopy_vol']].cov().values

# Calculate the PDF for each observation
pdf_values = multivariate_normal.pdf(df[['height', 'canopy_vol']], mean=mean_vector, cov=cov_matrix)
df['pdf'] = pdf_values

print("\nJoint Probability Density Function (PDF) Values:")
print(df[['height', 'canopy_vol', 'pdf']])


Joint Probability Density Function (PDF) Values:
    height  canopy_vol       pdf
0   17.997        0.71  0.450908
1   17.362        0.81  0.345655
2   18.148        0.92  0.119645
3   19.023        0.61  0.146158
4   17.266        0.73  0.518514
5   17.266        0.77  0.428861
6   19.079        0.85  0.058070
7   18.267        0.65  0.420170
8   17.031        0.98  0.081048
9   18.043        0.82  0.260223
10  17.037        0.41  0.647247
11  17.034        0.56  0.753683
12  17.742        0.59  0.685820
13  15.587        0.68  0.112237
14  15.775        0.69  0.152689
15  16.938        0.67  0.590964
16  16.487        0.60  0.516446
17  17.814        0.62  0.635965
18  16.592        0.53  0.605106
19  16.088        0.61  0.321132
20  18.966        0.11  0.015050
21  17.274        0.51  0.780819
22  17.568        0.13  0.113353
23  16.075        0.23  0.148540
24  16.956        0.35  0.519855
25  17.611        0.21  0.215058
26  16.349        0.15  0.110722
27  17.876        0.48  0.

**3.6e Marginal Probability**

In [11]:
# Marginal probability for height (integrate over canopy_vol)
marginal_height = df.groupby('height')['pdf'].sum()
print("\nMarginal Probability for Height:")
print(marginal_height)

# Marginal probability for canopy_vol (integrate over height)
marginal_canopy = df.groupby('canopy_vol')['pdf'].sum()
print("\nMarginal Probability for Canopy Volume:")
print(marginal_canopy)


Marginal Probability for Height:
height
15.540    0.147267
15.587    0.112237
15.775    0.152689
16.075    0.148540
16.088    0.321132
16.172    0.392989
16.279    0.335819
16.349    0.110722
16.442    0.503618
16.487    0.516446
16.592    0.605106
16.898    0.428566
16.899    0.408011
16.938    0.590964
16.956    0.519855
17.031    0.081048
17.034    0.753683
17.037    0.647247
17.208    0.492491
17.266    0.947375
17.274    0.780819
17.362    0.345655
17.487    0.765742
17.568    0.113353
17.611    0.215058
17.697    0.715210
17.709    0.694899
17.742    0.685820
17.814    0.635965
17.876    0.613488
17.997    0.450908
18.043    0.260223
18.148    0.119645
18.267    0.420170
18.323    0.348345
18.966    0.015050
19.023    0.146158
19.079    0.058070
19.352    0.069937
Name: pdf, dtype: float64

Marginal Probability for Canopy Volume:
canopy_vol
0.11    0.015050
0.13    0.113353
0.15    0.110722
0.21    0.215058
0.23    0.148540
0.30    0.408011
0.31    0.428566
0.33    0.492491
0.35

1.   Height 16.172: The marginal probability of observing a height of 16.172 is approximately 0.393. This means that when you consider all possible values of canopy_vol, the likelihood of encountering a height of 16.172 is about 39.3%.
2.   Height 18.966: The marginal probability of observing a height of 18.966 is very low (0.015). This indicates that such a height is quite rare in your dataset, with only about a 1.5% chance of occurrence.

**3.6f Conditional Probability**

In [13]:
# Conditional probability P(height | canopy_vol)
# Assuming height is the dependent variable

df['conditional_prob'] = df['pdf'] / df['canopy_vol'].map(marginal_canopy)

print("\nConditional Probability P(height | canopy_vol):")
print(df[['height', 'canopy_vol', 'conditional_prob']])


Conditional Probability P(height | canopy_vol):
    height  canopy_vol  conditional_prob
0   17.997        0.71          1.000000
1   17.362        0.81          1.000000
2   18.148        0.92          1.000000
3   19.023        0.61          0.312778
4   17.266        0.73          1.000000
5   17.266        0.77          1.000000
6   19.079        0.85          1.000000
7   18.267        0.65          1.000000
8   17.031        0.98          1.000000
9   18.043        0.82          1.000000
10  17.037        0.41          1.000000
11  17.034        0.56          0.496032
12  17.742        0.59          0.576591
13  15.587        0.68          1.000000
14  15.775        0.69          1.000000
15  16.938        0.67          1.000000
16  16.487        0.60          1.000000
17  17.814        0.62          1.000000
18  16.592        0.53          1.000000
19  16.088        0.61          0.687222
20  18.966        0.11          1.000000
21  17.274        0.51          0.917795
22  17.5

# MVND properties

# Import libraries

In [15]:
import pandas as pd
import numpy as np
from scipy.stats import multivariate_normal
from scipy.stats import shapiro
import pingouin as pg

***Now instead of taking only 2 varaible , we are taking 3 varaibles for Properties of MVND
#'height', 'canopy_vol', 'leaf_area'***

# Load data

In [17]:
# Define the data
data = {
    'plant_var': ['A']*10 + ['B']*10 + ['C']*10 + ['D']*10,
    'height': [
        17.997, 17.362, 18.148, 19.023, 17.266, 17.266, 19.079, 18.267, 17.031, 18.043,  # Group A
        17.037, 17.034, 17.742, 15.587, 15.775, 16.938, 16.487, 17.814, 16.592, 16.088,  # Group B
        18.966, 17.274, 17.568, 16.075, 16.956, 17.611, 16.349, 17.876, 16.899, 17.208,  # Group C
        16.898, 19.352, 17.487, 16.442, 18.323, 16.279, 17.709, 15.540, 16.172, 17.697   # Group D
    ],
    'canopy_vol': [
        0.71, 0.81, 0.92, 0.61, 0.73, 0.77, 0.85, 0.65, 0.98, 0.82,  # Group A
        0.41, 0.56, 0.59, 0.68, 0.69, 0.67, 0.60, 0.62, 0.53, 0.61,  # Group B
        0.11, 0.51, 0.13, 0.23, 0.35, 0.21, 0.15, 0.48, 0.30, 0.33,  # Group C
        0.31, 0.51, 0.56, 0.59, 0.42, 0.35, 0.50, 0.52, 0.55, 0.54   # Group D
    ],
    'leaf_area': [
        15.0, 16.5, 15.2, 14.8, 16.0, 15.7, 16.3, 15.8, 16.1, 15.6,  # Group A
        14.5, 14.7, 15.2, 14.3, 14.6, 14.9, 14.8, 15.3, 14.6, 14.9,  # Group B
        15.4, 15.1, 15.0, 14.8, 15.2, 15.3, 14.9, 15.1, 14.7, 15.0,  # Group C
        15.6, 15.8, 15.7, 15.9, 16.0, 15.6, 15.8, 15.7, 15.9, 16.1   # Group D
    ]
}
df2 = pd.DataFrame(data)

**Test Univariate Normality (Shapiro-Wilk Test)**

In [19]:
# Step: Test Univariate Normality (Shapiro-Wilk Test)
print("Testing Univariate Normality:")
for dep in ['height', 'canopy_vol', 'leaf_area']:
    stat, p = shapiro(df2[dep])
    print(f"Shapiro-Wilk Test for {dep}: W={stat:.3f}, p-value={p:.3f}")
    if p > 0.05:
        print(f"{dep} is normally distributed.\n")
    else:
        print(f"{dep} is NOT normally distributed.\n")

# Step: Test Multivariate Normality (Using Pingouin's multivariate_normality)
print("Testing Multivariate Normality:")
mv_norm_stat = pg.multivariate_normality(df2[['height', 'canopy_vol']], alpha=0.05)
print(mv_norm_stat)
print("\n")


Testing Univariate Normality:
Shapiro-Wilk Test for height: W=0.979, p-value=0.661
height is normally distributed.

Shapiro-Wilk Test for canopy_vol: W=0.980, p-value=0.695
canopy_vol is normally distributed.

Shapiro-Wilk Test for leaf_area: W=0.970, p-value=0.355
leaf_area is normally distributed.

Testing Multivariate Normality:
HZResults(hz=0.23529633435493413, pval=0.9575405933339186, normal=True)




**Create MVND matrix**

In [21]:
import pandas as pd
import numpy as np
from scipy.stats import multivariate_normal

# Calculate mean vector and covariance matrix for the entire dataset
mean_vector = df2[['height', 'canopy_vol', 'leaf_area']].mean().values
cov_matrix = df2[['height', 'canopy_vol', 'leaf_area']].cov().values

print("Mean Vector:\n", mean_vector)
print("\nCovariance Matrix:\n", cov_matrix)

# Define the multivariate normal distribution
mvn = multivariate_normal(mean=mean_vector, cov=cov_matrix)

Mean Vector:
 [17.281425  0.5365   15.335   ]

Covariance Matrix:
 [[0.9079103  0.02046358 0.18215397]
 [0.02046358 0.04552077 0.03579231]
 [0.18215397 0.03579231 0.30335897]]


# 3.6g Linear Combinations

In [23]:
# 1. Linear Combinations
# Define a 3x3 matrix of weights for different linear combinations
weights_matrix = np.array([
    [1, -0.5, 0.2],  # Weights for the first linear combination
    [0.5, 1, -0.3],  # Weights for the second linear combination
    [-0.2, 0.4, 1]   # Weights for the third linear combination
])

print("\nLinear Combinations:")
for i, weights in enumerate(weights_matrix):
    mean_linear_combination = np.dot(weights, mean_vector)
    variance_linear_combination = np.dot(weights, np.dot(cov_matrix, weights))

    print(f"Combination {i + 1}:")
    print(f"Mean of Linear Combination = {mean_linear_combination:.4f}")
    print(f"Variance of Linear Combination = {variance_linear_combination:.4f}\n")


Linear Combinations:
Combination 1:
Mean of Linear Combination = 20.0802
Variance of Linear Combination = 0.9767

Combination 2:
Mean of Linear Combination = 4.5767
Variance of Linear Combination = 0.2441

Combination 3:
Mean of Linear Combination = 12.0933
Variance of Linear Combination = 0.2995



Now genertae the the linear combination of weight matrix in order get tested whwether follow normal distribution or not. note: since the new datset will be the linear cobination of all 3 varaible so it has just one varaible data so we will apply uniaraite normal distribution test.

In [25]:
# Create data set
# Define the 3x3 matrix of weights
weights_matrix = np.array([
    [1, -0.5, 0.2],  # Weights for the first linear combination
    [0.5, 1, -0.3],  # Weights for the second linear combination
    [-0.2, 0.4, 1]   # Weights for the third linear combination
])

# Create new DataFrames based on the linear combinations
new_datasets = {}
for i, weights in enumerate(weights_matrix):
    # Apply weights to create a new variable
    new_data = df2[['height', 'canopy_vol', 'leaf_area']].dot(weights)
    # Create a new DataFrame with this new variable
    new_df = pd.DataFrame({
        'linear_combination': new_data
    })
    # Save to dictionary
    new_datasets[f'Combination_{i + 1}'] = new_df

# Print new datasets and check dimensions
for name, dataset in new_datasets.items():
    print(f"\n{name}:")
    print(dataset)  # Print entire DataFrame
    print(f"Shape of {name}: {dataset.shape}")  # Confirm the shape


Combination_1:
    linear_combination
0               20.642
1               20.257
2               20.728
3               21.678
4               20.101
5               20.021
6               21.914
7               21.102
8               19.761
9               20.753
10              19.732
11              19.694
12              20.487
13              18.107
14              18.350
15              19.583
16              19.147
17              20.564
18              19.247
19              18.763
20              21.991
21              20.039
22              20.503
23              18.920
24              19.821
25              20.566
26              19.254
27              20.656
28              19.689
29              20.043
30              19.863
31              22.257
32              20.347
33              19.327
34              21.313
35              19.224
36              20.619
37              18.420
38              19.077
39              20.647
Shape of Combination_1: (40, 1)

Combinat

**Make new dataframe from linear combination**

In [27]:
# Create dataframes for linear combinations
data_combination_1 = {
    'linear_combination': [
        20.642, 20.257, 20.728, 21.678, 20.101,
        20.021, 21.914, 21.102, 19.761, 20.753,
        19.732, 19.694, 20.487, 18.107, 18.350,
        19.583, 19.147, 20.564, 19.247, 18.763,
        21.991, 20.039, 20.503, 18.920, 19.821,
        20.566, 19.254, 20.656, 19.689, 20.043,
        19.863, 22.257, 20.347, 19.327, 21.313,
        19.224, 20.619, 18.420, 19.077, 20.647
    ]
}
new_df_combination_1 = pd.DataFrame(data_combination_1)

data_combination_2 = {
    'linear_combination': [
        5.2085, 4.5410, 5.4340, 5.6815, 4.5630,
        4.6930, 5.4995, 5.0435, 4.6655, 5.1615,
        4.5785, 4.6670, 4.9010, 4.1835, 4.1975,
        4.6690, 4.4035, 4.9370, 4.4460, 4.1840,
        4.9730, 4.6170, 4.4140, 3.8275, 4.2680,
        4.4255, 3.8545, 4.8880, 4.3395, 4.4340,
        4.0790, 5.4460, 4.5935, 4.0410, 4.7815,
        3.8095, 4.6145, 3.5800, 3.8660, 4.5585
    ]
}
new_df_combination_2 = pd.DataFrame(data_combination_2)

data_combination_3 = {
    'linear_combination': [
        11.6846, 13.3516, 11.9384, 11.2394, 12.8388,
        12.5548, 12.8242, 12.4066, 13.0858, 12.3194,
        11.2566, 11.5172, 11.8876, 11.4546, 11.7210,
        11.7804, 11.7426, 11.9852, 11.4936, 11.9264,
        11.6508, 11.8492, 11.5384, 11.6770, 11.9488,
        11.8618, 11.6902, 11.7168, 11.4402, 11.6904,
        12.3444, 12.1336, 12.4266, 12.8476, 12.5034,
        12.4842, 12.4582, 12.8000, 12.8856, 12.7766
    ]
}
new_df_combination_3 = pd.DataFrame(data_combination_3)

**TEST NORMALITY of new datframe**

In [29]:
from scipy.stats import shapiro

# List of DataFrames and their names
dfs = [new_df_combination_1, new_df_combination_2, new_df_combination_3]
names = ["Combination 1", "Combination 2", "Combination 3"]

print("Testing Univariate Normality:")
for df, name in zip(dfs, names):
    stat, p = shapiro(df['linear_combination'])
    print(f"{name}: Shapiro-Wilk Test W={stat:.3f}, p-value={p:.3f}")
    if p > 0.05:
        print(f"{name} is normally distributed.\n")
    else:
        print(f"{name} is NOT normally distributed.\n")

Testing Univariate Normality:
Combination 1: Shapiro-Wilk Test W=0.981, p-value=0.738
Combination 1 is normally distributed.

Combination 2: Shapiro-Wilk Test W=0.980, p-value=0.692
Combination 2 is normally distributed.

Combination 3: Shapiro-Wilk Test W=0.943, p-value=0.043
Combination 3 is NOT normally distributed.



# 3.6h Subset Distribution

In [31]:
# 2. Subset Distribution
# Take a subset: height and canopy_vol
subset_mean = mean_vector[:2]  # Mean for height and canopy_vol
subset_cov = cov_matrix[:2, :2]  # Covariance matrix for height and canopy_vol

print("\nSubset Distribution:")
print(f"Mean of Subset = {subset_mean}")
print(f"Covariance Matrix of Subset =\n{subset_cov}")


Subset Distribution:
Mean of Subset = [17.281425  0.5365  ]
Covariance Matrix of Subset =
[[0.9079103  0.02046358]
 [0.02046358 0.04552077]]


**CONDUCT MVND on this new subset data**

In [33]:
# Define the data
data = {
    'plant_var': ['A']*10 + ['B']*10 + ['C']*10 + ['D']*10,
    'height': [
        17.997, 17.362, 18.148, 19.023, 17.266, 17.266, 19.079, 18.267, 17.031, 18.043,  # Group A
        17.037, 17.034, 17.742, 15.587, 15.775, 16.938, 16.487, 17.814, 16.592, 16.088,  # Group B
        18.966, 17.274, 17.568, 16.075, 16.956, 17.611, 16.349, 17.876, 16.899, 17.208,  # Group C
        16.898, 19.352, 17.487, 16.442, 18.323, 16.279, 17.709, 15.540, 16.172, 17.697   # Group D
    ],
    'canopy_vol': [
        0.71, 0.81, 0.92, 0.61, 0.73, 0.77, 0.85, 0.65, 0.98, 0.82,  # Group A
        0.41, 0.56, 0.59, 0.68, 0.69, 0.67, 0.60, 0.62, 0.53, 0.61,  # Group B
        0.11, 0.51, 0.13, 0.23, 0.35, 0.21, 0.15, 0.48, 0.30, 0.33,  # Group C
        0.31, 0.51, 0.56, 0.59, 0.42, 0.35, 0.50, 0.52, 0.55, 0.54   # Group D
    ]
}
dfsub = pd.DataFrame(data)

#Test Univariate Normality (Shapiro-Wilk Test)
print("Testing Univariate Normality:")
for dep in ['height', 'canopy_vol']:
    stat, p = shapiro(dfsub[dep])
    print(f"Shapiro-Wilk Test for {dep}: W={stat:.3f}, p-value={p:.3f}")
    if p > 0.05:
        print(f"{dep} is normally distributed.\n")
    else:
        print(f"{dep} is NOT normally distributed.\n")

# Test Multivariate Normality (Using Pingouin's multivariate_normality)
print("Testing Multivariate Normality:")
mv_norm_stat = pg.multivariate_normality(dfsub[['height', 'canopy_vol']], alpha=0.05)
print(mv_norm_stat)
print("\n")

Testing Univariate Normality:
Shapiro-Wilk Test for height: W=0.979, p-value=0.661
height is normally distributed.

Shapiro-Wilk Test for canopy_vol: W=0.980, p-value=0.695
canopy_vol is normally distributed.

Testing Multivariate Normality:
HZResults(hz=0.23529633435493413, pval=0.9575405933339186, normal=True)




# 3.6i Zero Covariance and Independence

In [35]:
# Check covariance between height and canopy_vol
cov_height_canopy = cov_matrix[0, 1]
print("\nCovariance between height and canopy_vol:")
print(cov_height_canopy)

# Zero covariance implies independence if variables are normally distributed
# For demonstration: calculate pairwise covariance
print("\nCovariance Matrix:")
print(cov_matrix)

# Check if any off-diagonal elements are close to zero
is_independent = np.allclose(cov_matrix[0, 1], 0)
print("\nIndependence Check:")
print(f"Height and Canopy Volume are independent: {is_independent}")


Covariance between height and canopy_vol:
0.02046357692307684

Covariance Matrix:
[[0.9079103  0.02046358 0.18215397]
 [0.02046358 0.04552077 0.03579231]
 [0.18215397 0.03579231 0.30335897]]

Independence Check:
Height and Canopy Volume are independent: False


In [37]:
# Assuming df2 is your DataFrame with the relevant columns

# Calculate mean vector and covariance matrix for the entire dataset
mean_vector = df2[['height', 'canopy_vol', 'leaf_area']].mean().values
cov_matrix = df2[['height', 'canopy_vol', 'leaf_area']].cov().values

print("Mean Vector:\n", mean_vector)
print("\nCovariance Matrix:\n", cov_matrix)

# Define the multivariate normal distribution
mvn = multivariate_normal(mean=mean_vector, cov=cov_matrix)

# Variable names for reference
variable_names = ['Height', 'Canopy Volume', 'Leaf Area']

# Iterate through each pair of variables
for i in range(len(variable_names)):
    for j in range(i + 1, len(variable_names)):
        # Extract covariance between variable i and variable j
        cov_ij = cov_matrix[i, j]

        # Check for zero covariance
        is_independent = np.allclose(cov_ij, 0)

        # Print results
        print(f"\nCovariance between {variable_names[i]} and {variable_names[j]}:")
        print(cov_ij)

        print(f"Independence Check for {variable_names[i]} and {variable_names[j]}:")
        print(f"{variable_names[i]} and {variable_names[j]} are independent: {is_independent}")

# Add variance information for completeness
print("\nVariances:")
for i in range(len(variable_names)):
    var = cov_matrix[i, i]
    print(f"{variable_names[i]} variance: {var}")

Mean Vector:
 [17.281425  0.5365   15.335   ]

Covariance Matrix:
 [[0.9079103  0.02046358 0.18215397]
 [0.02046358 0.04552077 0.03579231]
 [0.18215397 0.03579231 0.30335897]]

Covariance between Height and Canopy Volume:
0.02046357692307684
Independence Check for Height and Canopy Volume:
Height and Canopy Volume are independent: False

Covariance between Height and Leaf Area:
0.18215397435897443
Independence Check for Height and Leaf Area:
Height and Leaf Area are independent: False

Covariance between Canopy Volume and Leaf Area:
0.03579230769230771
Independence Check for Canopy Volume and Leaf Area:
Canopy Volume and Leaf Area are independent: False

Variances:
Height variance: 0.9079103019230775
Canopy Volume variance: 0.045520769230769205
Leaf Area variance: 0.3033589743589744


# 3.6j Conditoinonal distribution of componets are MVN

In [39]:
# Define the data
data = {
    'plant_var': ['A']*10 + ['B']*10 + ['C']*10 + ['D']*10,
    'height': [
        17.997, 17.362, 18.148, 19.023, 17.266, 17.266, 19.079, 18.267, 17.031, 18.043,  # Group A
        17.037, 17.034, 17.742, 15.587, 15.775, 16.938, 16.487, 17.814, 16.592, 16.088,  # Group B
        18.966, 17.274, 17.568, 16.075, 16.956, 17.611, 16.349, 17.876, 16.899, 17.208,  # Group C
        16.898, 19.352, 17.487, 16.442, 18.323, 16.279, 17.709, 15.540, 16.172, 17.697   # Group D
    ],
    'canopy_vol': [
        0.71, 0.81, 0.92, 0.61, 0.73, 0.77, 0.85, 0.65, 0.98, 0.82,  # Group A
        0.41, 0.56, 0.59, 0.68, 0.69, 0.67, 0.60, 0.62, 0.53, 0.61,  # Group B
        0.11, 0.51, 0.13, 0.23, 0.35, 0.21, 0.15, 0.48, 0.30, 0.33,  # Group C
        0.31, 0.51, 0.56, 0.59, 0.42, 0.35, 0.50, 0.52, 0.55, 0.54   # Group D
    ],
    'leaf_area': [
        15.0, 16.5, 15.2, 14.8, 16.0, 15.7, 16.3, 15.8, 16.1, 15.6,  # Group A
        14.5, 14.7, 15.2, 14.3, 14.6, 14.9, 14.8, 15.3, 14.6, 14.9,  # Group B
        15.4, 15.1, 15.0, 14.8, 15.2, 15.3, 14.9, 15.1, 14.7, 15.0,  # Group C
        15.6, 15.8, 15.7, 15.9, 16.0, 15.6, 15.8, 15.7, 15.9, 16.1   # Group D
    ]
}
df2 = pd.DataFrame(data)


In [ ]:
# Calculate mean vector and covariance matrix for the entire dataset
mean_vector = df2[['height', 'canopy_vol', 'leaf_area']].mean().values
cov_matrix = df2[['height', 'canopy_vol', 'leaf_area']].cov().values

# Define the multivariate normal distribution
mvn = multivariate_normal(mean=mean_vector, cov=cov_matrix)

# Indices for the variables

In [ ]:
index_height = 0
index_canopy_vol = 1
index_leaf_area = 2

# Extract the covariance matrix components
cov_matrix_11 = cov_matrix[index_height, index_height]
cov_matrix_12 = cov_matrix[index_height, index_canopy_vol]
cov_matrix_13 = cov_matrix[index_height, index_leaf_area]
cov_matrix_22 = cov_matrix[index_canopy_vol, index_canopy_vol]
cov_matrix_23 = cov_matrix[index_canopy_vol, index_leaf_area]
cov_matrix_33 = cov_matrix[index_leaf_area, index_leaf_area]

# Partition the covariance matrix
cov_x1_x1 = cov_matrix_11
cov_x1_x2 = np.array([cov_matrix_12, cov_matrix_13])
cov_x2_x2 = cov_matrix_22 * np.identity(2) + np.array([[cov_matrix_12, cov_matrix_23], [cov_matrix_23, cov_matrix_33]])

# Compute conditional distribution parameters
mean_x1_given_x2 = mean_vector[index_height] + cov_x1_x2 @ np.linalg.inv(cov_x2_x2) 
@ (mean_vector[[index_canopy_vol, index_leaf_area]] - mean_vector[[index_canopy_vol, index_leaf_area]])
cov_x1_given_x2 = cov_x1_x1 - cov_x1_x2 @ np.linalg.inv(cov_x2_x2) @ cov_x1_x2.T

print("Conditional Mean of Height given Canopy Volume and Leaf Area:\n", mean_x1_given_x2)
print("\nConditional Covariance Matrix of Height given Canopy Volume and Leaf Area:\n", cov_x1_given_x2)

# Optional: Define conditional distribution
conditional_mvn = multivariate_normal(mean=mean_x1_given_x2, cov=cov_x1_given_x2)

# 3.6k Maximum Likelihood Estimation (MLE) for Mean and Covariance Matrix

In [43]:
import pandas as pd
import numpy as np

# Assuming df2 is your DataFrame with the relevant columns
data = df2[['height', 'canopy_vol', 'leaf_area']].values

# Compute the MLE for the mean vector
mean_mle = np.mean(data, axis=0)

# Compute the MLE for the covariance matrix
cov_matrix_mle = np.cov(data, rowvar=False)

print("MLE for Mean Vector:\n", mean_mle)
print("\nMLE for Covariance Matrix:\n", cov_matrix_mle)

MLE for Mean Vector:
 [17.281425  0.5365   15.335   ]

MLE for Covariance Matrix:
 [[0.9079103  0.02046358 0.18215397]
 [0.02046358 0.04552077 0.03579231]
 [0.18215397 0.03579231 0.30335897]]


# 3.6l Wishart Distribution

# Import libraries

In [45]:
import pandas as pd
import numpy as np
from scipy.stats import wishart

# Load data

In [47]:
# Assuming df2 is your DataFrame with the relevant columns
data = df2[['height', 'canopy_vol', 'leaf_area']].values

# Wishart Distribution

In [49]:
# Compute the sample covariance matrix from df2
cov_matrix_mle = np.cov(data, rowvar=False)

# Number of variables (dimensions)
p = cov_matrix_mle.shape[0]

# Degrees of freedom for Wishart distribution (should be >= p)
df = p + 2  # Example value; typically df > p

# Define the Wishart distribution with the sample covariance matrix
wishart_dist = wishart(df=df, scale=cov_matrix_mle)

# Generate a sample from the Wishart distribution
wishart_sample = wishart_dist.rvs(size=1)

print("Sample from the Wishart Distribution:\n", wishart_sample)

# Compute the mean and covariance matrix of the Wishart sample (for verification)
mean_wishart = np.mean(wishart_sample, axis=0)
cov_wishart = np.cov(wishart_sample[0], rowvar=False)

print("Mean of Wishart Sample:\n", mean_wishart)
print("\nCovariance Matrix of Wishart Sample:\n", cov_wishart)

Sample from the Wishart Distribution:
 [[ 7.4834472  -0.88583012  2.16274539]
 [-0.88583012  0.23744867  0.03186919]
 [ 2.16274539  0.03186919  1.63460697]]
Mean of Wishart Sample:
 [ 2.92012083 -0.20550409  1.27640719]

Covariance Matrix of Wishart Sample:
 17.941413870547386
